proposals/NNNN-service-context.md

```diff
@@ -0,0 +1,177 @@
+
+One can create a service context using the `.topLevel` factory method which intends to clarify that this should not be used UNLESS intending to start a fresh "top level" context that does not inherit any values from the current context:
+
+```swift
+var context = ServiceContext.topLevel
+```
+
+In most situations, developers should prefer to "pick up" the context from the current context:
+
+```swift
+var context: ServiceContext? = ServiceContext.current
+```
+
+which inspects the task-local values for the presence of a service context. It is by design that it is possible to determine if no context was set, or if an empty context is set. When intending to "carry existing, or create a new" context, the following pattern is used:
+
+```swift
+var context = ServiceContext.current ?? ServiceContext.topLevel
+```
+
+Once obtained, one can set values in by using defined keys, like this:
+
+```swift
+private enum FirstTestKey: ServiceContextKey {
+  typealias Value = Int
+}
+
+var context = ServiceContext.topLevel
+context[FirstTestKey.self] = 42
+```
+
+Keys declared as types conforming to `ServiceContextKey` allow for future extension where we might want to configure specific keys using additional behavior, like for example defining the `static let nameOverride` of a key. It is also important that a Key type may be private to whomever declares it, allowing only such library to set these values. 
+
+Service context is used primarily as a task-local value read and modified like this:
+
+```swift
+func exampleFunction() async -> Int {
+  guard let context = ServiceContext.current {
+    return 0
+  }
+  guard let value = context[FirstTestKey.self] {
+    return 0
+  }
+  print("test = \(value)") // test = 42
+  return value
+}
+
+// ----------------------------------------
+
+var context = ServiceContext.topLevel
+context[FirstTestKey.self] = 42
+
+let c = ServiceContext.withValue(context) {
+    await exampleFunction()
+}
+assert(c == 42)
+```
+
+Swift's [task local values](https://developer.apple.com/documentation/swift/tasklocal) are used to automatically propagate the context value to any child tasks that may be created from this code, such that context is automatically propagated through to them, e.g. like this:
+
+```swift
+func testMeMore() {
+  guard let context = ServiceContext.current {
+    return 0
+  }
+  guard let value = context[FirstTestKey.self] {
+    return 0
+  }
+  return value
+}
+
+func test() async -> Int {
+  async let v = testMeMore() // read context from child task
+  return await v
+}
+
+var context = ServiceContext.topLevel
+context[FirstTestKey.self] = 42
+ServiceContext.withValue(context) {
+    assert(test() == 42)
+}
+```
+
+For keys which are expected to be public API, used ty developers, it is recommended to follow this pattern to declare an accessor on the context, rather than exposing the Key type directly:
+
+```swift
+extension ServiceContext { 
```

Comment:
Shouldn't we recommend the nice pattern for this

```
extension ServiceContext {
    private enum TestKey: ServiceContextKey {
        typealias Value = Int
    }
    public var testValue: String? {
         set { 
              self[TestKey.self] = newValue
         }
          get {
              self[TestKey.self]
         }
}
```

That way, there's one way to access and we don't leak the somewhat unituitive type.

Reply:
That is what this is trying to show, yeah. Agreed that the snippet should include the type, will fix

Reply:
I polished this all up a bit so we're strictly recommending that pattern